### DISCUSSION TITLE
Use SimpleMeshLayer and MVT

### DISCUSSION THREAD

**Discussion by am2222** [visgl/deck.gl#9173] [Q&A] [answered]
Hi, 
Is it possible to combine SimpleMeshLayer and MVT layers?

I am using something like this
```js
new MVTLayer({
      binary: true,
      id: 'mvt-layer',
      data: `http://0.0.0.0:3000/{z}/{x}/{y}?time=${time.toISOString()}`,
     

      minZoom: 9,
      maxZoom: 20,
      uniqueIdProperty: 'id',
      pickable: true,
      autoHighlight: true,
      renderSubLayers: props => {
        return [new GeoJsonLayer({
          ...props,
          iconAtlas: '/icon.png',
          iconMapping: '/icon_config.json',
          getIcon: d => 'icon',
          getIconSize: (d) => 25,
          getIconAngle: (d) => 30,
          iconSizeUnits: 'pixels',
          iconSizeScale: 1,
          autoHighlight: true,
          pointType: 'icon',
          },
        }),
      
        new SimpleMeshLayer({
          ...props,
          id: 'SimpleMeshLayer',
          getColor: (d) => [Math.sqrt(d.exits), 140, 0],
          getOrientation: (d) => [0, Math.random() * 180, 0],
          getPosition: (d) => d.coordinates,
          mesh: 'https://raw.githubusercontent.com/visgl/deck.gl-data/master/website/humanoid_quad.obj',
          sizeScale: 30,
          pickable: true,
          loaders: [OBJLoader]
        })
      ]
      }


    })

```
However I get the following error
```
deck.js:82 deck: initialization of SimpleMeshLayer({id: 'SimpleMeshLayer'}): deck.gl: assertion failed. Error: deck.gl: assertion failed.
```

but it does not give a clear explanation of what properties are causing the issue

**Comment by ibgreen** [comment]
> but it does not give a clear explanation of what properties are causing the issue

As a principle, we do not include messages regarding programming mistakes when throwing exceptions, since all those strings would add considerable to the bundle size of the library.

However we often include comments in the source code where we assert, so that developers can use the "break on exceptions" setting in the browser development tools to break on the exception and figure out what is going wrong that way.

Can't guarantee that is the case here, but worth a try to see if you can spot the error that way.

**Reply by am2222** [reply]
Thanks @ibgreen !
so I have been debugging my code and I ended up seeing that it fails in this line.


https://github.com/visgl/deck.gl/blob/9067f45a7d9424e397bc823ad89f3e010ddf1b31/modules/core/src/lib/layer.ts#L884


This assertion fails if I have a `SimpleMeshLayer` as sublayer. 

Here I tried to reproduce it. I tried to use either `SimpleMeshLayer` or `ScenegraphLayer` and they both faild 

https://codepen.io/Am3333/pen/qBeBpLd?editors=1111

**Reply by ibgreen** [reply]
Most likely cause is that you need to make sure each mesh layer has a unique `id` property.

**Reply by am2222** [reply]
Thanks @ibgreen 
so I modified my code to make sure the id is dynamic however I am still getting the same error 
```
deck: initialization of ScenegraphLayer({id: 'mvt-layer-648-1583-12'}): deck.gl: assertion failed. Error: deck.gl: assertion failed.
    at assert6 (assert.js:6:15)
    at _ScenegraphLayer._initialize (layer.js:680:9)
    at LayerManager._initializeLayer (layer-manager.js:287:19)
```
The same line is failing. I have modified the codepen too

**Comment by grahambates-code** [comment, marked as answer]
I think its a data issue, I don't know your data properly, but this renders

https://codepen.io/grahambates/pen/mdNyGxV?editors=1111

**Reply by am2222** [reply]
Thanks @grahambates-code 

May I ask if the code you shared renders 3d? For me it does not show anything when I open it!

**Reply by am2222** [reply]
Thanks @grahambates-code it works for me now! I'd why it was not showing anything before!
I think the part I was missing was to convert the coordinates manually.
```
 const positions = props.data?.points?.positions?.value || [];
          const pairs = [];
          for (let i = 0; i < positions.length; i += 2) {
            pairs.push([positions[i], positions[i + 1]]);
          }
```
